File: rental/src/main/java/io/axoniq/demo/bikerental/rental/handler/CommandHandlerController.java
package io.axoniq.demo.bikerental.rental.handler;

import io.axoniq.demo.bikerental.rental.command.BikeCommandHandler;
import io.axoniq.demo.bikerental.rental.support.AxonServerMessages;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;
import reactor.core.publisher.Mono;

import java.util.Map;
import java.util.concurrent.CompletionException;

/**
 * Receives commands that Axon Server routes to this application over the Integration HTTP API (the
 * endpoint registered by {@code IntegrationRegistrar}). It hands the command to the
 * {@link BikeCommandHandler} and returns a command result; a failure is reported as a 500 so Axon
 * Server relays it to the original sender.
 */
@RestController
public class CommandHandlerController {

    private final BikeCommandHandler commandHandler;

    public CommandHandlerController(BikeCommandHandler commandHandler) {
        this.commandHandler = commandHandler;
    }

    @PostMapping("/axon/command")
    public Mono<ResponseEntity<Object>> handle(@RequestBody AxonServerMessages.Command command) {
        return Mono.fromFuture(commandHandler.handle(command))
                   .map(payload -> ok(command.id(), payload))
                   .defaultIfEmpty(ok(command.id(), null))
                   .onErrorResume(error -> Mono.just(failure(error)));
    }

    private static ResponseEntity<Object> ok(String id, com.fasterxml.jackson.databind.JsonNode payload) {
        return ResponseEntity.ok(new AxonServerMessages.CommandResult(id, null, null, null, payload));
    }

    private static ResponseEntity<Object> failure(Throwable error) {
        Throwable cause = error instanceof CompletionException && error.getCause() != null
                ? error.getCause() : error;
        return ResponseEntity.status(500).body(Map.of("message", String.valueOf(cause.getMessage())));
    }
}

File: rental/src/handler/command-handler-controller.ts
import { Router } from 'express';
import type { Command, CommandResult } from '@bikerental/core-api';
import { BikeCommandHandler } from '../command/bike-command-handler.js';

/**
 * Receives commands that Axon Server routes to this application over the Integration HTTP API. It
 * hands the command to the {@link BikeCommandHandler} and returns a command result; a failure is
 * reported as a 500 so Axon Server relays it to the original sender.
 */
export function commandHandlerRouter(commandHandler: BikeCommandHandler): Router {
  const router = Router();
  router.post('/axon/command', async (req, res) => {
    const command = req.body as Command;
    try {
      const payload = await commandHandler.handle(command);
      const result: CommandResult = { id: command.id, payload: payload ?? null };
      res.json(result);
    } catch (error) {
      res.status(500).json({ message: String((error as Error)?.message) });
    }
  });
  return router;
}
